Objects/genobject.c

```diff
@@ -273,6 +273,17 @@ gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
     return result ? PYGEN_RETURN : PYGEN_ERROR;
 }
 
+static inline PySendResult
+gen_send_ex2(PyGenObject *gen, PyObject *arg, PyObject **presult,
+             int exc, int closing)
+{
+    PySendResult result;
+    Py_BEGIN_CRITICAL_SECTION(gen);
```

Comment:
If we start to put locks, maybe other functions would deserve that lock as well:

* _PyGen_yf()
* gen_close()
* _gen_throw()
* _gen_getframe()

What about these getters?

* gen_getrunning()
* gen_getsuspended()

In frameobject.c, see frame_is_suspended() (should be safe).

In ceval.c, see clear_gen_frame().

Reply:
With some stress testing, it looks mostly ok, but there is a race with `gi_frame_state`. I'll play around with fixes, we might be able to just use `_Py_atomic` instead of a critical section, which would improve performance.